### PR TITLE
endpoint: Fix packets to host dropped with the chaining mode and host firewall

### LINKED ISSUE #18942: Packets to host dropped with chaining mode and host firewall
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Cilium drops packets destined to host when hostFirewall is enabled here
https://github.com/cilium/cilium/blob/b8506a1dcc8ccc1eb07117953a46f7c6181742e3/bpf/bpf_lxc.c#L788
https://github.com/cilium/cilium/blob/b8506a1dcc8ccc1eb07117953a46f7c6181742e3/bpf/bpf_lxc.c#L1745

```
$ kubectl -n kube-system exec cilium-n9vcc -- cilium monitor --type drop
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), clean-cilium-state (init)
Listening for events on 12 CPUs with 64x4096 of shared memory
Press Ctrl-C to quit
level=info msg="Initializing dissection cache..." subsys=monitor
xx drop (Missed tail call) flow 0x7bda48fd to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58314 tcp SYN, ACK
xx drop (Missed tail call) flow 0x36a2383e to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58536 tcp SYN, ACK
xx drop (Missed tail call) flow 0x0 to endpoint 0, , identity 19387->host: 172.19.0.1:3128 -> 10.69.1.133:43088 tcp ACK, RST
xx drop (Missed tail call) flow 0x1356d934 to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58516 tcp SYN, ACK
xx drop (Missed tail call) flow 0x6f751770 to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58440 tcp SYN, ACK
xx drop (Missed tail call) flow 0x51ffc6cc to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58482 tcp SYN, ACK
xx drop (Missed tail call) flow 0xd56deee0 to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58324 tcp SYN, ACK
xx drop (Missed tail call) flow 0xecfdc48d to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58326 tcp SYN, ACK
xx drop (Missed tail call) flow 0x7f2930f to endpoint 0, , identity 19387->host: 172.19.0.1:53 -> 10.69.1.133:58550 tcp SYN, ACK

$ kubectl -n kube-system exec cilium-n9vcc -- bpftool map dump pinned /sys/fs/bpf/tc/globals/cilium_call_policy
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), clean-cilium-state (init)
key: 26 03 00 00  value: c0 0f 00 00
key: ec 03 00 00  value: c4 0f 00 00
key: bd 0a 00 00  value: 4a 11 00 00
key: 0e 0c 00 00  value: 4c 11 00 00
Found 4 elements
```

helm values
```
devices: "eth+ eno1+ eno2+"
cni:
  chainingMode: "generic-veth"
  customConf: true
datapathMode: "veth"
tunnel: "disabled"
enableIPv4Masquerade: false
enableIdentityMark: false
hostFirewall:
  enabled: true
policyEnforcementMode: "default"
policyAuditMode: true
kubeProxyReplacement: "disabled"
prometheus:
  enabled: true
rollOutCiliumPods: true
resources:
  requests:
    cpu: 100m
    memory: 400Mi
operator:
  rollOutPods: true
  prometheus:
    enabled: true
  resources:
    requests:
      cpu: 100m
      memory: 200Mi
hubble:
  relay:
    enabled: true
    tls:
      server:
        enabled: true
    rollOutPods: true
    resources:
      requests:
        cpu: 100m
        memory: 200Mi
  tls:
    auto:
      method: "cronJob"
  metrics:
    enabled:
      - "dns"
      - "drop:destinationContext=pod|dns|ip;sourceContext=pod|dns|ip"
      - "tcp"
      - "flow:destinationContext=pod|dns|ip;sourceContext=pod|dns|ip"
      - "icmp"
      - "http"
```

### Cilium Version

```
$ kubectl -n kube-system exec cilium-n9vcc -- cilium version
Defaulted container "cilium-agent" out of: cilium-agent, mount-cgroup (init), clean-cilium-state (init)
Client: 1.11.0 27e0848cc5 2021-12-05T15:34:41-08:00 go version go1.17.3 linux/amd64
Daemon: 1.11.0 27e0848cc5 2021-12-05T15:34:41-08:00 go version go1.17.3 linux/amd64
```

### Kernel Version

```
$ uname -a
Linux rack2-cs5 5.10.96-flatcar #1 SMP Thu Feb 3 20:00:07 -00 2022 x86_64 AMD EPYC 7413 24-Core Processor AuthenticAMD GNU/Linux
```

### Kubernetes Version

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"clean", BuildDate:"2021-12-16T08:38:33Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"archive", BuildDate:"2021-12-20T06:57:14Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
```

### Sysdump

_No response_

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ysksuzuki**
This PR fixes the issue Cilium drops packets to the host when it's configured with the chaining mode and host firewall by populating the policy map regardless of the per-endpoint routes setting. All routing is performed by the Linux stack with the chaining mode even if the per-endpoint route is disabled in the agent.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #18942

### CODE REVIEW

**Comment by kkourt** [comment]
Added @pchaigno as a reviewer, since it seems that he has some context.

File: pkg/datapath/loader/template.go [hunk 1]
```diff
@@ -167,10 +167,10 @@ func elfMapSubstitutions(ep datapath.Endpoint) map[string]string {
 		}
 	}
 
-	// The policy map is only used for the host endpoint if per-endpoint
-	// routes and the host firewall are enabled.
-	if !ep.IsHost() ||
-		(option.Config.EnableEndpointRoutes && option.Config.EnableHostFirewall) {
+	// Populate the policy map if the host firewall is enabled regardless of the per-endpoint route setting
```
Comment (pchaigno) [inline pkg/datapath/loader/template.go:170]: Nit: This should be wrap like other lines. Not sure why checkpatch doesn't complain. cc @qmonnet

**Review by pchaigno — APPROVED** [review]
LGTM! Only one small nit, probably not worth repushing unless some other changes are needed.

File: pkg/datapath/loader/template.go (hunk 1 above)
Comment (qmonnet) [reply pkg/datapath/loader/template.go:170]: Because it's Go code and we're not passing it through checkpatch.pl

Comment (pchaigno) [reply pkg/datapath/loader/template.go:170]: Ah, of course! :facepalm:

**Comment by ysksuzuki** [comment]
It seems that some tests are failing. I don't see those related to this change. Should I rebase my branch?

https://jenkins.cilium.io/job/Cilium-PR-K8s-1.21-kernel-5.4/1296/testReport/Suite-k8s-1/21/K8sUpdates_Tests_upgrade_and_downgrade_from_a_Cilium_stable_image_to_master/

```
Stderr:
 	 WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: /home/jenkins/workspace/Cilium-PR-K8s-1.21-kernel-5.4/src/github.com/cilium/cilium/test/vagrant-kubeconfig
	 WARNING: Kubernetes configuration file is world-readable. This is insecure. Location: /home/jenkins/workspace/Cilium-PR-K8s-1.21-kernel-5.4/src/github.com/cilium/cilium/test/vagrant-kubeconfig
	 Error: INSTALLATION FAILED: rendered manifests contain a resource that already exists. Unable to continue with install: Service "hubble-peer" in namespace "kube-system" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "cilium-preflight": current value is "cilium"
	 

/home/jenkins/workspace/Cilium-PR-K8s-1.21-kernel-5.4/src/github.com/cilium/cilium/test/k8s/updates.go:127
```

**Comment by pchaigno** [comment]
> Should I rebase my branch?

@ysksuzuki Yes, please. The K8sUpdates test failures should be fixed by https://github.com/cilium/cilium/pull/19759.

**Comment by ysksuzuki** [comment]
@pchaigno Rebased my branch. Please rerun the tests

**Comment by ysksuzuki** [comment]
Something wrong with the test environment. I can't imagine that my patch would affect it.

https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-kernel-4.9/1596/testReport/junit/Suite-k8s-1/16/K8sPolicyTest_Multi_node_policy_test_with_L7_policy_using_connectivity_check_to_check_datapath/

```
Stderr:
 	 The Service "echo-b" is invalid: spec.ports[0].nodePort: Invalid value: 31414: provided port is already allocated
```

**Comment by pchaigno** [comment]
k8s-1.16-kernel-4.9 failed with known flake https://github.com/cilium/cilium/issues/13071. Other required tests are passing.
We're missing a review from @cilium/endpoint but I've reviewed and am the author of the specific changed lines under @cilium/endpoint's ownership so fine to ignore the missing review IMO.
Marking ready to merge.

**Review by christarazi — APPROVED** [review]

**Pull request merged**